[CL-2216] Display fixed_ratio_layout banner

alexander-cit opened the pull request:
Inspired by https://github.com/CitizenLabDotCo/citizenlab/pull/3536

alexander-cit committed:
[CL-2216] Display fixed_ratio_layout banner

alexander-cit commented on front/app/components/LandingPages/citizen/HeaderContent.ts:
```diff
@@ -54,12 +54,9 @@ export const HeaderTitle = styled.h1<{
-
   ${media.phone`
-    margin-bottom: 15px;
+    font-size: ${fontSizes.xl}px;
```
Before:
![image](https://user-images.githubusercontent.com/91883434/207948883-879cbbd8-543d-49d6-ab60-ff36bbb4c1ac.png)

After:
![image](https://user-images.githubusercontent.com/91883434/207948931-bf237315-6e59-457d-9ca9-2afe7edf1872.png)

alexander-cit commented on front/app/components/LandingPages/citizen/HeaderContent.ts:
```diff
@@ -88,6 +85,12 @@ export const HeaderSubtitle = styled.h2<{
   margin: 0;
   margin-bottom: 15px;
 
+  ${media.phone`
```
After:
![image](https://user-images.githubusercontent.com/91883434/207950464-9fd216b3-b20f-4233-863a-1da1d412622e.png)

brentguf replied:
We need more margin-bottom on the sub-header text. The button is too close to it.

It should be similar to when we have the avatar bubbles: if the avatar bubbles are the last element before the button, they have more margin-bottom. Add a similar margin bottom to the subheader text if there are no avatar bubbles.

alexander-cit replied:
I suppose this issue wasn't introduced in this PR, was it? In this case, could we do it in a separate PR?

We already have logic for it, but it works only for the  `displayHeaderAvatar` setting https://github.com/CitizenLabDotCo/citizenlab/blob/b816374a55d9dfd3e6be81d0ea6e42cd024a116e/front/app/components/LandingPages/citizen/HeaderContent.ts#L91-L94
I'm not sure how to make it work with the conditions inside of the `AvatarBubbles` component.

alexander-cit committed:
[CL-2216] Display fixed_ratio_layout banner on custom page

alexander-cit committed:
[CL-2216] Import width constants from components library

alexander-cit committed:
[CL-2216] Rename some components. Remove some unneeded code

alexander-cit committed:
[CL-2216] Get rid of duplicated import

brentguf committed:
Merge branch 'master' of https://github.com/CitizenLabDotCo/citizenlab into CL-2216-display-fixed_ratio_layout

brentguf committed:
Merge branch 'CL-1849/image-dimensions' of https://github.com/CitizenLabDotCo/citizenlab into CL-2216-display-fixed_ratio_layout

brentguf committed:
Remove unused import

brentguf committed:
Merge branch 'CL-1849/image-dimensions' of https://github.com/CitizenLabDotCo/citizenlab into CL-2216-display-fixed_ratio_layout

brentguf reviewed (changes requested):
The banner needs some work on mobile. If we have avatar bubbles, there's not enough space:

<img width="374" alt="Screenshot 2022-12-19 at 11 27 50" src="https://user-images.githubusercontent.com/16427929/208405495-149c6d11-53a5-4b4c-bede-15c80fff14f0.png">
<img width="621" alt="Screenshot 2022-12-19 at 11 27 28" src="https://user-images.githubusercontent.com/16427929/208405510-b50b75cc-e04c-43c6-9047-8a6440a31302.png">

If there's no other option, not showing the avatar bubbles on mobile can be the solution.

If you have questions or concerns, bring it to #dev-flexible-pages so the others can be involved as well.

alexander-cit committed:
[CL-2216] Make banner look fine with avatar bubbles

alexander-cit commented:
@brentguf 
![image](https://user-images.githubusercontent.com/91883434/208711041-7d2195a0-33ad-4208-b929-92b540f90008.png)
![image](https://user-images.githubusercontent.com/91883434/208711168-98ef88c4-e0c5-4862-a772-af9f5c2bbf97.png)

https://github.com/CitizenLabDotCo/citizenlab/pull/3494/commits/49202c6999e051dd704cbfe237e11cb1460e976e (probably, I should put some styles in better places, but the idea is like that)

brentguf committed:
Merge branch 'CL-1849/image-dimensions' of https://github.com/CitizenLabDotCo/citizenlab into CL-2216-display-fixed_ratio_layout

brentguf commented:
@alexander-cit let's pair on this if you're stuck.

alexander-cit commented:
@brentguf It's done, you can see the screenshots.

brentguf commented:
> @brentguf It's done, you can see the screenshots.

There's not enough room above the text/below the button.

One way to solve: hide avatar bubbles if fixed-ratio layout on mobile. You could check with Yaesul too.

(Unassigned myself for now.)

IvaKop commented:
@alexander-cit Let's add some basic unit tests for the display. As a general rule, we should add unit tests when we create new components, unless there's a really good reason not to. Checking if the code is tested should also become a bigger part of our PR process. Will add it to the CDM discussion points for this week.

alexander-cit commented:
@IvaKop

It just seems that this PR only has CSS and plain HTML. The little logic is covered by type checks. So, the test will duplicate existing markup and will fail only if the requirements change. Then we'll need to fix it. So, it will just introduce double work and no protection from bugs.

Let me know if you've already considered it. In this case, I'll gladly add the tests.

IvaKop commented:
You should definitely not test the markup only but there's quite some display logic in there, as well (when should a component appear and when not). Will mark some examples to give you an idea.

IvaKop commented on front/app/containers/CustomPageShow/CustomPageHeader/index.tsx:
```diff
@@ -21,6 +22,9 @@ const CustomPageHeader = ({ pageData }: Props) => {
       {pageAttributes.banner_layout === 'two_row_layout' && (
         <TwoRowLayout pageData={pageData} />
       )}
+      {pageAttributes.banner_layout === 'fixed_ratio_layout' && (
```
Unit test - does this component show when it should. Bonus, if we check the other layouts, as well.

IvaKop commented on front/app/containers/HomePage/SignedOutHeader/FixedRatioLayout.tsx:
```diff
@@ -0,0 +1,46 @@
+  const homepageSettings = useHomepageSettings();
+
+  if (!isNilOrError(homepageSettings)) {
+    const headerImage = homepageSettings.attributes.header_bg?.large;
```
Does this work as expected - displaying the component, the image and background as it should based on the BE data?

IvaKop commented on front/app/containers/HomePage/SignedOutHeader/HeaderContent.tsx:
```diff
@@ -102,7 +102,9 @@ const HeaderContent = ({
         {displayHeaderAvatars && <StyledAvatarBubbles />}
 
-        <CTA signedIn={false} buttonStyle={buttonStyle} signUpIn={signUpIn} />
+        <div style={{ marginBottom: '20px' }}>
```
Not related to tests but let's avoid using inline styles in this way. We have established mechanisms - a `Box` component that provides this functionality and we can always use a style component, if something more is needed.

IvaKop commented on front/app/containers/HomePage/SignedOutHeader/index.tsx:
```diff
@@ -22,6 +23,7 @@ const SignedOutHeaderIndex = () => {
         )}
         {homepageBannerLayout === 'two_column_layout' && <TwoColumnLayout />}
         {homepageBannerLayout === 'two_row_layout' && <TwoRowLayout />}
+        {homepageBannerLayout === 'fixed_ratio_layout' && <FixedRatioLayout />}
```
We could also add a test for this logic.

alexander-cit committed:
Revert "[CL-2216] Make banner look fine with avatar bubbles"

This reverts commit 49202c6999e051dd704cbfe237e11cb1460e976e.

alexander-cit committed:
Merge branch 'CL-1849/image-dimensions' into CL-2216-display-fixed_ratio_layout

alexander-cit committed:
[CL-2216] Change banner height depending on content

alexander-cit committed:
[CL-2216] Fixes after merge

alexander-cit committed:
[CL-2216] Add unit tests for fixed-ratio

alexander-cit commented:
@IvaKop I've added some tests https://github.com/CitizenLabDotCo/citizenlab/pull/3494/commits/131f8b41d718eb7ea8ca4a5382b5b5b552ad40df

But to my taste, they look quite brittle and pretty useless. I would gladly pair on it tomorrow if you know how we can improve them.

alexander-cit committed:
[CL-2216] Rename index.test.tsx files to ComponentName.test.tsx

alexander-cit committed:
[CL-2216] Add more tests

alexander-cit committed:
[CL-2216] Fix tests failing with CSS.supports is not a function

This CSS.supports check is used only for IE, which is not supported
anymore.

The error was:
in Image (created by TwoColumnLayout__HeaderImage)
detail: TypeError: CSS.supports is not a function
  at Image.supports [as render] (/home/main/projects/cl/citizenlab/front/app/components/UI/Image/index.tsx:121:41)

It could be reproduced by running:
npm test app/containers/CustomPageShow/CustomPageHeader/CustomPageHeader.test.tsx

alexander-cit committed:
[CL-2216] Add more tests for HomePage/SignedOutHeader

brentguf commented on front/app/components/LandingPages/citizen/FixedRatioBannerLayout.ts:
```diff
@@ -0,0 +1,27 @@
+export const Header = styled.div`
+  width: 100%;
+  max-width: ${stylingConsts.pageWidth}px;
+  min-height: 225px;
```
Could you take this from homepageBannerLayoutHeights like the other layouts?

brentguf replied:
Does this banner require a min-height in any case?

alexander-cit committed:
Revert "[CL-2216] Change banner height depending on content"

This reverts commit 0822aef142d619831f758ae78b05654822cd272c.

alexander-cit committed:
[CL-2216] Hide avatars on mobile with fixed-ratio layout
